esdt-safe/src/to_sovereign/create_tx.rs

```diff
@@ -112,114 +115,205 @@ pub trait CreateTxModule:
+    }
+
+    fn check_and_extract_fee(
+        &self,
+    ) -> MultiValue2<OptionalValue<EsdtTokenPayment>, ManagedVec<EsdtTokenPayment>> {
         let mut payments = self.call_value().all_esdt_transfers().clone_value();
-        let fees_payment = self.pop_first_payment(&mut payments);
+        let fee_market_address = self.fee_market_address().get();
+        let fee_enabled_mapper = SingleValueMapper::new_from_address(
+            fee_market_address.clone(),
+            StorageKey::from("feeEnabledFlag"),
+        )
+        .get();
+
+        let opt_transfer_data = if fee_enabled_mapper {
+            OptionalValue::Some(self.pop_first_payment(&mut payments))
+        } else {
+            OptionalValue::None
+        };
 
-        require!(!payments.is_empty(), "Nothing to transfer");
-        require!(payments.len() <= MAX_TRANSFERS_PER_TX, "Too many tokens");
+        MultiValue2::from((opt_transfer_data, payments))
+    }
 
-        let opt_gas_limit = match &opt_transfer_data {
+    fn process_transfer_data(
+        &self,
+        opt_transfer_data: OptionalValue<
+            MultiValue3<GasLimit, ManagedBuffer, ManagedVec<ManagedBuffer>>,
+        >,
+    ) -> Option<TransferData<Self::Api>> {
+        match &opt_transfer_data {
             OptionalValue::Some(transfer_data) => {
+                let (gas_limit, function, args) = transfer_data.clone().into_tuple();
                 let max_gas_limit = self.max_user_tx_gas_limit().get();
-                require!(
-                    transfer_data.gas_limit <= max_gas_limit,
-                    "Gas limit too high"
-                );
+
+                require!(gas_limit <= max_gas_limit, "Gas limit too high");
 
                 require!(
-                    !self
-                        .banned_endpoint_names()
-                        .contains(&transfer_data.function),
+                    !self.banned_endpoint_names().contains(&function),
                     "Banned endpoint name"
                 );
 
-                OptionalValue::Some(transfer_data.gas_limit)
+                Some(TransferData {
+                    gas_limit,
+                    function,
+                    args,
+                })
             }
-            OptionalValue::None => OptionalValue::None,
-        };
+            OptionalValue::None => None,
+        }
+    }
+
+    /// Create an Elrond -> Sovereign transaction.
+    #[payable("*")]
+    #[endpoint]
+    fn deposit(
+        &self,
+        to: ManagedAddress,
+        opt_transfer_data: OptionalValue<
+            MultiValue3<GasLimit, ManagedBuffer, ManagedVec<ManagedBuffer>>,
+        >,
+    ) {
+        require!(self.not_paused(), "Cannot create transaction while paused");
+
+        let (fees_payment, payments) = self.check_and_extract_fee().into_tuple();
+
+        require!(!payments.is_empty(), "Nothing to transfer");
+        require!(payments.len() <= MAX_TRANSFERS_PER_TX, "Too many tokens");
```

Comment:
you can put these inside `check_and_extract_fee`

Reply:
done